File: src/main/java/com/xpdustry/nucleus/NucleusPlugin.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus;

import com.google.gson.Gson;
import com.xpdustry.foundation.annotation.PluginAnnotationProcessor;
import com.xpdustry.foundation.plugin.BaseMindustryPlugin;
import com.xpdustry.foundation.plugin.PluginFacade;
import com.xpdustry.foundation.plugin.PluginListener;
import com.xpdustry.foundation.plugin.PluginLogger;
import com.xpdustry.nucleus.concurrent.AsyncAwareScheduledTaskHandlerProcessor;
import com.xpdustry.nucleus.concurrent.ConcurrentModule;
import com.xpdustry.nucleus.config.ConfigModule;
import com.xpdustry.nucleus.database.DatabaseModule;
import com.xpdustry.nucleus.dependency.DependencyModule;
import com.xpdustry.nucleus.dependency.DependencyService;
import com.xpdustry.nucleus.gatekeeper.GatekeeperModule;
import com.xpdustry.nucleus.message.MessageModule;
import com.xpdustry.nucleus.metric.MetricCollectorHandlerProcessor;
import com.xpdustry.nucleus.metric.MetricModule;
import com.xpdustry.nucleus.network.NetworkModule;
import com.xpdustry.nucleus.text.TextModule;
import java.net.http.HttpClient;
import java.nio.file.Path;

public final class NucleusPlugin extends BaseMindustryPlugin {

    private final DependencyService dependencies = DependencyService.create(
            new PluginModule(),
            new ConfigModule(),
            new DatabaseModule(),
            new MessageModule(),
            new TextModule(),
            new NetworkModule(),
            new MetricModule(),
            new GatekeeperModule(),
            new ConcurrentModule());

    @Override
    public void onInit() {
        for (final var dependency : this.dependencies.resolveAll()) {
            if (dependency instanceof PluginListener listener && !(dependency instanceof PluginFacade)) {
                this.addListener(listener);
            }
        }

        final var processors = PluginAnnotationProcessor.compose(
                PluginAnnotationProcessor.events(this),
                PluginAnnotationProcessor.triggers(this),
                PluginAnnotationProcessor.playerActions(this),
                this.dependencies.resolve(AsyncAwareScheduledTaskHandlerProcessor.class),
                this.dependencies.resolve(MetricCollectorHandlerProcessor.class));
        for (final var listener : this.listeners()) {
            processors.process(listener);
        }
    }

    private final class PluginModule implements DependencyModule {

        @Override
        public void configure(final DependencyService.Binder binder) {
            final var directory = NucleusPlugin.this.directory();
            binder.bindInstance(Path.class, "config", directory.resolve("config.properties"));
            binder.bindInstance(Path.class, "postgres", directory.resolve("postgres"));
            binder.bindInstance(PluginLogger.class, NucleusPlugin.this.logger());
            binder.bindInstance(PluginFacade.class, NucleusPlugin.this);
            binder.bindInstance(Gson.class, new Gson());
            binder.bindInstance(HttpClient.class, HttpClient.newHttpClient());
        }
    }
}

File: src/main/kotlin/com/xpdustry/nucleus/NucleusPlugin.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus;

import com.google.gson.Gson;
import com.xpdustry.foundation.annotation.PluginAnnotationProcessor;
import com.xpdustry.foundation.plugin.BaseMindustryPlugin;
import com.xpdustry.nucleus.config.ConfigManager;
import com.xpdustry.nucleus.database.PostgresDatabase;
import com.xpdustry.nucleus.gatekeeper.GatekeeperController;
import com.xpdustry.nucleus.gatekeeper.GatekeeperPipeline;
import com.xpdustry.nucleus.message.MessagePublisher;
import com.xpdustry.nucleus.metric.MetricExporter;
import com.xpdustry.nucleus.metric.MindustryMetricCollector;
import com.xpdustry.nucleus.network.InetAddressInfoProvider;
import com.xpdustry.nucleus.network.InetAddressWhitelist;
import com.xpdustry.nucleus.text.BadWordFinder;
import java.net.http.HttpClient;

public final class NucleusPlugin extends BaseMindustryPlugin {

    private final PluginAnnotationProcessor<?> processor = PluginAnnotationProcessor.compose(
            PluginAnnotationProcessor.events(this),
            PluginAnnotationProcessor.scheduledTasks(this),
            PluginAnnotationProcessor.playerActions(this));

    private final ConfigManager configManager =
            this.addListener(new ConfigManager(this.directory().resolve("config.properties")));

    private final Gson gson = new Gson();
    private final HttpClient httpClient = HttpClient.newHttpClient();

    private final PostgresDatabase database = this.addListener(
            new PostgresDatabase(this.configManager, this.directory().resolve("postgres")));

    private final MessagePublisher network = this.addListener(new MessagePublisher(this.configManager, this.database));

    private final BadWordFinder badWords = this.addListener(new BadWordFinder(this.gson));

    private final InetAddressInfoProvider addressInfoProvider = this.addListener(
            new InetAddressInfoProvider(this.configManager, this.gson, this.httpClient, this.database));
    private final InetAddressWhitelist addressWhitelist = new InetAddressWhitelist(this.database);

    private final GatekeeperPipeline gatekeeperPipeline = new GatekeeperPipeline();

    private final MetricExporter metrics = this.addListener(new MetricExporter(this.configManager, this.httpClient));

    @Override
    public void onInit() {
        this.metrics.register(this.addListener(new MindustryMetricCollector()), false);
        this.addListener(new GatekeeperController(
                this.gatekeeperPipeline,
                this.configManager,
                this.badWords,
                this.addressInfoProvider,
                this.addressWhitelist));
        for (final var listener : this.listeners()) {
            this.processor.process(listener);
        }
    }
}
